File: api/server-common/src/main/java/su/plo/lib/api/server/chat/ServerTextConverter.java
package su.plo.lib.api.server.chat;

import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import su.plo.lib.api.chat.MinecraftTextComponent;
import su.plo.lib.api.chat.MinecraftTranslatableText;
import su.plo.lib.api.chat.TranslatableTextConverter;
import su.plo.lib.api.server.command.MinecraftCommandSource;
import su.plo.voice.api.server.config.ServerLanguages;

import java.util.Map;
import java.util.function.Supplier;

@RequiredArgsConstructor
public abstract class ServerTextConverter<T> extends TranslatableTextConverter<T> {

    private final Supplier<ServerLanguages> languagesSupplier;

    public T convert(@NotNull MinecraftCommandSource source, @NotNull MinecraftTextComponent text) {
        ServerLanguages languages = languagesSupplier.get();
        if (languages == null || !(text instanceof MinecraftTranslatableText))
            return convert(text);

        Map<String, String> language = languages.getServerLanguage(source);
        MinecraftTranslatableText translatable = translateArguments(language, (MinecraftTranslatableText) text);

        if (!language.containsKey(translatable.getKey()))
            return convert(text);

        return convert(translate(language, translatable));
    }
}

File: api/server-common/src/main/kotlin/su/plo/lib/api/server/chat/ServerTextConverter.kt
package su.plo.lib.api.server.chat

import su.plo.lib.api.chat.MinecraftTextComponent
import su.plo.lib.api.chat.MinecraftTranslatableText
import su.plo.lib.api.chat.TranslatableTextConverter
import su.plo.lib.api.server.command.MinecraftCommandSource
import su.plo.voice.api.server.config.ServerLanguages
import java.util.function.Supplier

abstract class ServerTextConverter<T> constructor(
    private val languagesSupplier: Supplier<ServerLanguages?>
) : TranslatableTextConverter<T>() {

    fun convertToJson(source: MinecraftCommandSource, text: MinecraftTextComponent): String {
        val languages = languagesSupplier.get() ?: return convertToJson(text)
        val language = languages.getServerLanguage(source)

        val convertedText = translateInner(language, text)
        if (convertedText !is MinecraftTranslatableText)
            return convertToJson(text)

        if (!language.containsKey(convertedText.key))
            return convertToJson(text)

        return convertToJson(translate(
            language,
            convertedText
        ))
    }

    fun convert(source: MinecraftCommandSource, text: MinecraftTextComponent): T {
        val languages = languagesSupplier.get() ?: return convert(text)
        val language = languages.getServerLanguage(source)

        val convertedText = translateInner(language, text)
        if (convertedText !is MinecraftTranslatableText)
            return convert(text)

        if (!language.containsKey(convertedText.key))
            return convert(text)

        return convert(translate(
            language,
            convertedText
        ))
    }
}
